### ISSUE TITLE
[SPIRV] Empty struct in cbuffer causes compilation to fail

### ISSUE THREAD

**Issue by baldurk** [microsoft/DirectXShaderCompiler#2882]
If an empty struct is declared in a cbuffer (or a nested struct inside) the compilation seems to break trying to align members properly to their alignment requirements around that struct:

```
struct empty_struct { };

cbuffer cb
{
	float3 a;
	empty_struct b;
	float2 c;

	float4 test;
};

float4 main() : SV_TARGET
{
	return test;
}
```

When compiling with latest dxc (`dxcompiler.dll: 1.6 - 1.5.0.2653 (f0d4419f)`) I get this error:

```
fatal error: generated SPIR-V is invalid: Structure id 4 decorated as Block for variable in Uniform storage class must follow relaxed uniform buffer layout rules: member 1 at offset 12 is not aligned to 16
  %type_cb = OpTypeStruct %v3float %empty_struct %v2float %v4float

note: please file a bug report on https://github.com/Microsoft/DirectXShaderCompiler/issues with source code if possible
```

With the offset of 12 coming wrongly from the `float2 c` trying to pack directly after `float3 a`. If I allow scalar layout with `-fvk-use-scalar-layout` then it doesn't complain since a tight packing is valid.

**Comment by ehsannas** [comment]
Thanks for reporting @baldurk

**Comment by jeremy-lunarg** [comment]
This is an old issue, but I took a look at it. The empty struct is being assigned an offset of 12 by DXC. Relevant spir-v:
```
    OpMemberName %type_cb 0 "a"
    OpMemberName %type_cb 1 "b"
    OpMemberName %type_cb 2 "c"
    OpMemberName %type_cb 3 "test"
    OpMemberDecorate %type_cb 0 Offset 0
    OpMemberDecorate %type_cb 1 Offset 12
    OpMemberDecorate %type_cb 2 Offset 16
    OpMemberDecorate %type_cb 3 Offset 32
```

For reference glslang produces the following offsets:

```
    OpMemberDecorate %cb 0 Offset 0
    OpMemberDecorate %cb 1 Offset 16
    OpMemberDecorate %cb 2 Offset 16
    OpMemberDecorate %cb 3 Offset 32
```

That seems reasonable. Using `-fvk-use-scalar-layout` in DXC also produces reasonable offsets:

```
    OpMemberDecorate %type_cb 0 Offset 0
    OpMemberDecorate %type_cb 1 Offset 12
    OpMemberDecorate %type_cb 2 Offset 12
    OpMemberDecorate %type_cb 3 Offset 20
```

However, while investigating the spir-v spec, I noticed the following validation rule regarding composite objects:

"- The ArrayStride, MatrixStride, and Offset decorations must be large enough to hold the size of
the objects they affect (that is, specifying overlap is invalid). Each ArrayStride and MatrixStride
must be greater than zero, and **it is invalid for two members of a given structure to be assigned the
same Offset.**"

Am I misinterpreting the spec?
